packages/smoldot-provider/src/SmoldotProvider.ts

```diff
@@ -0,0 +1,352 @@
+        chain_spec: this.#chainSpec,
+        json_rpc_callback: (response: string) => {
+            this.#handleRpcReponse(response);
+        },
+        database_save_callback: (database_content: string) => { 
+          l.debug('saving database', database_content);
+          this.#db.save(database_content);
+        }
+      })
+      .then((client: smoldot.SmoldotClient) => {
+        this.#client = client;
+        this.#isConnected = true;
+        this.emit('connected');
+      })
+      .catch((error: Error) => {
+        console.error(error);
+      });
+  }
+
+  /**
+   * @description Manually "disconnect" - drops the reference to the WASM client
+   */
+  // eslint-disable-next-line @typescript-eslint/require-await
+  public async disconnect(): Promise<void> {
+    if (this.#client) {
+      this.#client = undefined;
+    }
+    this.#isConnected = false;
+    this.emit('disconnected');
+  }
+
+  /**
+   * @summary Whether the node is connected or not.
+   * @return {boolean} true if connected
+   */
+  public get isConnected (): boolean {
+    return this.#isConnected;
+  }
+
+  /**
+   * @summary Listen to provider events - in practice the smoldot provider only
+   * emits a `connected` event after successfully starting the smoldot client
+   * and `disconnected` after `disconnect` is called.
+   * @param type - Event
+   * @param sub - Callback
+   */
+  public on(
+    type: ProviderInterfaceEmitted,
+    sub: ProviderInterfaceEmitCb
+  ): () => void {
+    this.#eventemitter.on(type, sub);
+
+    return (): void => {
+      this.#eventemitter.removeListener(type, sub);
+    };
+  }
+
+  /**
+   * @summary Send JSON data using WebSockets to the wasm node.
+   * @param method The RPC methods to execute
+   * @param params Encoded paramaters as appliucable for the method
+   * @param subscription Subscription details (internally used by `subscribe`)
+   */
+  public async send(
+    method: string,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    params: any[],
+    subscription?: SubscriptionHandler
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  ): Promise<any> {
+    return new Promise((resolve, reject): void => {
+        assert(this.isConnected && this.#client, 'Client is not connected');
+
+        const json = this.#coder.encodeJson(method, params);
+        const id = this.#coder.getId();
+
+        const callback = (error?: Error | null, result?: any): void => {
+          error
+            ? reject(error)
+            : resolve(result);
+        };
+
+        l.debug(() => ['calling', method, json]);
+
+        this.#handlers[id] = {
+          callback,
+          method,
+          params,
+          subscription
+        };
+
+      // Should we really catch errors here?  What can we do if the wasm client
+      // throws? Are we in an inconsistent state?
+      try {
+        this.#client.send_json_rpc(json);
+      } catch (error) {
+        reject(error);
+      }
```

Comment:
It's not supposed to throw except under catastrophic situations. I'd suggest to simply not deal with the error.